src/Compilers/VisualBasic/Portable/Binding/InitializerSemanticModel.vb

```diff
@@ -134,7 +134,20 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End Select
 
             Dim expressionInitializer = TryCast(boundInitializer, BoundExpression)
-            Return If(expressionInitializer IsNot Nothing, New BoundEqualsValue(initializer, expressionInitializer), boundInitializer)
+            If expressionInitializer IsNot Nothing Then
+                Select Case Me.MemberSymbol.Kind
+                    Case SymbolKind.Parameter
+                        Return New BoundParameterEqualsValue(initializer, DirectCast(Me.RootBinder.ContainingMember, ParameterSymbol), expressionInitializer)
+                    Case SymbolKind.Property
+                        Return New BoundPropertyInitializer(initializer, ImmutableArray.Create(DirectCast(Me.MemberSymbol, PropertySymbol)), Nothing, expressionInitializer)
+                    Case SymbolKind.Field
+                        Return New BoundFieldInitializer(initializer, ImmutableArray.Create(DirectCast(Me.MemberSymbol, FieldSymbol)), Nothing, expressionInitializer)
+                    Case Else
+                        Throw ExceptionUtilities.Unreachable
+                End Select
+            End If
```

Comment:
The TryCast can't be avoided, but the second switch and the additional DirectCasts can.